v3/mathjax3-ts/core/MathDocument.ts

```diff
@@ -0,0 +1,401 @@
+     *  for the document.
+     *
+     * @param{OptionList} options  The options for locating the math
+     * @return{MathDocument}       The math document instance
+     */
+    findMath(options?: OptionList): MathDocument;
+
+    /*
+     * Calls the input jax to process the MathItems in the MathList
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    compile(): MathDocument;
+
+    /*
+     * Gets the metric information for the MathItems
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    getMetrics(): MathDocument;
+
+    /*
+     * Calls the output jax to process the compiled math in the MathList
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    typeset(): MathDocument;
+
+    /*
+     * Add any event handlers to the typeset math
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    addEventHandlers(): MathDocument;
+
+    /*
+     * Updates the document to include the typeset math
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    updateDocument(): MathDocument;
+
+    /*
+     * Removes the typeset math from the document
+     *
+     * @param{boolean} restore  True if the original math should be put
+     *                            back into the document as well
+     * @return{MathDocument}    The math document instance
+     */
+    removeFromDocument(restore?: boolean): MathDocument;
+
+    /*
+     * Set the state of the document (allowing you to roll back
+     *  the state to a previous one, if needed).
+     *
+     * @param{boolean} restore  True if the original math should be put
+     *                            back into the document during the rollback
+     * @return{MathDocument}    The math document instance
+     */
+    state(state: number, restore?: boolean): MathDocument;
+
+    /*
+     * Clear the processed values so that the document can be reprocessed
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    reset(): MathDocument;
+
+    /*
+     * Reset the processed values and clear the MathList (so that new math
+     * can be processed in the document).
+     *
+     * @return{MathDocument}  The math document instance
+     */
+    clear(): MathDocument;
+
+    /*
+     * Merges a MathList into the list for this document.
+     *
+     * @param{MathList} list   The MathList to be merged into this document's list
+     * @return{MathDocument}   The math document instance
+     */
+    concat(list: MathList): MathDocument;
+
+}
+
+/*****************************************************************/
+/*
+ *  The MathDocument interface
+ */
+
+export interface MathDocumentClass {
+    /*
+     * The kind of MathDocument (e.g., "HTML")
+     */
+    KIND: string;
+
+    /*
+     * The default options for this type of document
+     */
+    OPTIONS: OptionList;
+
+    /*
+     *  The document states
+     */
+    STATE: {[name: string]: number};
```

Comment:
The possible set of states is not a fixed list.  It may be augmented by the subclasses of MathDocument, as needed.